# Fine-tuning Gemma 3 4B for Docker CLI Command Translation

This notebook trains a 4B parameter model to translate natural language
to Docker CLI commands. Runs on free Colab T4 GPU in ~45-60 minutes.

**What you'll build:**
- 85-95% accuracy Docker command translator
- ~2.5GB quantized model
- Runs locally on CPU (~3-5s inference)

**No ML experience required** - all steps explained.

## Configuration

Change these values to train for a different CLI tool:
```python
CLI_TOOL = "docker"
DATASET_FILE = "data/docker_training.jsonl"
OUTPUT_PREFIX = "docker_gemma3_4b"
```

In [1]:
!git clone https://github.com/pranavkumaarofficial/nlcli-wizard.git


Cloning into 'nlcli-wizard'...
remote: Enumerating objects: 134, done.
remote: Counting objects: 100% (33/33), done.
remote: Compressing objects: 100% (27/27), done.
remote: Total 134 (delta 6), reused 23 (delta 4), pack-reused 101 (from 1)
Receiving objects: 100% (134/134), 77.46 MiB | 13.21 MiB/s, done.
Resolving deltas: 100% (53/53), done.
Encountered 1 file(s) that should have been pointers, but weren't:
	models/readme.md


In [2]:
# Go to your cloned repo directory first
%cd /content/nlcli-wizard

# Pull the latest changes from the remote branch
!git pull origin main -r --autostash
!apt-get install git-lfs -y
!git lfs install
!git lfs pull




import torch

print("CUDA available:", torch.cuda.is_available())
print("Device:", torch.device("cuda" if torch.cuda.is_available() else "cpu"))
print("GPU name:", torch.cuda.get_device_name(0) if torch.cuda.is_available() else "No GPU")



/content/nlcli-wizard
From https://github.com/pranavkumaarofficial/nlcli-wizard
 * branch            main       -> FETCH_HEAD
Already up to date.
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
git-lfs is already the newest version (3.0.2-1ubuntu0.3).
0 upgraded, 0 newly installed, 0 to remove and 41 not upgraded.
Updated git hooks.
Git LFS initialized.
CUDA available: True
Device: cuda
GPU name: Tesla T4


# Fine-tuning Gemma 3 4B for Docker CLI Translation

REPO: https://github.com/pranavkumaarofficial/nlcli-wizard

**Project**: nlcli-wizard  
**Model**: google/gemma-3-4b-it  
**Technique**: QLoRA with Unsloth (Dynamic 4-bit)  
**Hardware**: Google Colab T4 GPU (Free Tier)  
**CLI Tool**: Docker

---

## Why 4B over 1B?

After 3 training runs with Gemma 3 1B, we hit a **capacity ceiling at 73-76%** accuracy.
The 1B model's 13M trainable parameters cannot hold all Docker flag patterns simultaneously —
fixing one category (e.g., exec `-it`) causes regressions in others (run, compose).

The 4B model provides ~4x more capacity, which should break through this ceiling.

## What You'll Learn

1. **Why Gemma 3 4B?** - 4x capacity over 1B, still fits on free T4 GPU
2. **What is Unsloth?** - How it makes training 2x faster with 70% less VRAM
3. **QLoRA Explained** - Low-rank adaptation for efficient fine-tuning
4. **4-bit Quantization** - How to compress models without losing accuracy
5. **Dynamic Quantization** - Unsloth's smart approach to preserving critical weights
6. **GGUF Format** - Converting for CPU inference with llama.cpp

---

## Training Objective

Fine-tune Gemma 3 4B to translate natural language → Docker CLI commands:

```
Input:  "run nginx on port 8080 in background"
Output: "docker run -d -p 8080:80 nginx"
```

**Target Accuracy**: 85-95% on domain-specific commands (up from 73-76% with 1B)

---

# Step 1: Setup and Installation

## Install Unsloth and Dependencies

### What is Unsloth?

**Unsloth** is a highly optimized library for fine-tuning LLMs that provides:

- **2x Faster Training**: Custom CUDA kernels optimized for LoRA operations
- **70% Less VRAM**: Efficient memory management and gradient checkpointing
- **Dynamic 4-bit Quantization**: Smart weight selection (don't quantize critical layers)
- **Zero Accuracy Loss**: Maintains full precision where it matters

### How Unsloth Works:

```
Traditional Fine-tuning:
├── Load full model (FP16) → 8GB VRAM for 4B
├── Compute gradients for ALL parameters
└── Update all 4B parameters → SLOW

Unsloth + QLoRA:
├── Load model in 4-bit → ~2.5GB VRAM
├── Add small LoRA adapters (~50MB)
├── Compute gradients ONLY for adapters → FAST
└── Update ~1.3% of parameters → 2x speed, 70% less VRAM
```

### Dynamic 4-bit Quantization:

Unsloth analyzes your model and **selectively avoids quantizing** critical layers:
- Attention output layers
- Layer norms
- Embedding layers

Result: **10% more VRAM but significantly better accuracy**

In [3]:
# Install Unsloth with all optimizations
# This will take ~3-5 minutes on first run

%%capture
!pip install "unsloth[colab-new] @ git+https://github.com/unslothai/unsloth.git"
!pip install --no-deps xformers trl peft accelerate bitsandbytes

print("✅ Unsloth and dependencies installed!")

In [4]:
# Verify GPU is available
import torch

if torch.cuda.is_available():
    gpu_name = torch.cuda.get_device_name(0)
    gpu_memory = torch.cuda.get_device_properties(0).total_memory / 1e9
    print(f"✅ GPU Available: {gpu_name}")
    print(f"   Total VRAM: {gpu_memory:.1f} GB")
else:
    print("❌ No GPU detected! This notebook requires a GPU.")
    print("   Go to Runtime → Change runtime type → Select T4 GPU")

✅ GPU Available: Tesla T4
   Total VRAM: 15.8 GB


---

# Step 2: Clone Repository and Load Dataset


In [5]:
!git clone https://github.com/pranavkumaarofficial/nlcli-wizard.git
import os
# Change to project directory
os.chdir('/content/nlcli-wizard')

print("\n✅ Repository cloned successfully!")
print(f"   Current directory: {os.getcwd()}")

Cloning into 'nlcli-wizard'...
remote: Enumerating objects: 134, done.
remote: Counting objects: 100% (33/33), done.
remote: Compressing objects: 100% (27/27), done.
remote: Total 134 (delta 6), reused 23 (delta 4), pack-reused 101 (from 1)
Receiving objects: 100% (134/134), 77.46 MiB | 15.53 MiB/s, done.
Resolving deltas: 100% (53/53), done.
Updating files: 100% (117/117), done.
Encountered 1 file(s) that should have been pointers, but weren't:
	models/readme.md

✅ Repository cloned successfully!
   Current directory: /content/nlcli-wizard


In [ ]:
# === CONFIGURATION ===
# Change these to train for a different CLI tool
CLI_TOOL = "docker"
DATASET_FILE = "data/docker_training.jsonl"
OUTPUT_PREFIX = "docker_gemma3_4b"

# Verify dataset exists and inspect it
import json
from pathlib import Path

dataset_path = Path(DATASET_FILE)

if not dataset_path.exists():
    print(f"Dataset not found at {DATASET_FILE}!")
    print("Make sure you pushed the latest code to GitHub")
else:
    # Load and inspect dataset
    examples = []
    with open(dataset_path, 'r') as f:
        for line in f:
            examples.append(json.loads(line))

    print(f"Dataset loaded: {len(examples)} examples")
    print(f"CLI Tool: {CLI_TOOL}")
    print(f"\nSample Examples:")
    print("-" * 80)

    for i, ex in enumerate(examples[:3]):
        print(f"\nExample {i+1}:")
        print(f"  Instruction: {ex['instruction']}")
        print(f"  Output: {ex['output'].strip()}")

    print("-" * 80)

---

# Step 3: Load Gemma 3 4B with Unsloth

## Understanding Model Loading

### What happens when we load a model?

1. **Download from HuggingFace** (~8GB for Gemma 3 4B in FP16)
2. **Load into GPU memory** with quantization
3. **Prepare for training** with LoRA adapters

### Quantization Explained:

**Normal Precision (FP16)**:
```
Weight: 0.123456789 (16 bits) → 2 bytes per parameter
4B parameters × 2 bytes = 8 GB
```

**4-bit Quantization (NF4)**:
```
Weight: 0.123456789 → Quantized to 4 bits (0-15)
4B parameters × 0.5 bytes = 2 GB
```

**NF4 (Normal Float 4-bit)**:
- Special quantization format optimized for neural network weights
- Weights follow normal distribution, so use non-uniform quantization
- More precision for common values, less for outliers

### Dynamic 4-bit:

Unsloth's smart feature:
```python
if layer_is_critical():  # Attention, embeddings, norms
    keep_fp16()  # Don't quantize
else:
    quantize_4bit()  # Safe to compress
```

Result: **~2.5GB VRAM** (instead of 8GB) with minimal accuracy loss

### T4 GPU Memory Budget (15.8 GB):
```
Model (4-bit):     ~2.5 GB
LoRA adapters:     ~50 MB
Optimizer states:  ~1.5 GB
Activations/grad:  ~4-6 GB
Overhead:          ~2 GB
Total:             ~10-12 GB ✅ Fits on T4!
```

In [ ]:
from unsloth import FastLanguageModel
import torch

# Model configuration
model_name = "unsloth/gemma-3-4b-it"  # Unsloth's optimized 4B version
max_seq_length = 512  # Maximum context length for our task
dtype = None  # Auto-detect (FP16 for T4 GPU)
load_in_4bit = True  # Enable 4-bit quantization

print("Loading Gemma 3 4B with Unsloth optimizations...")
print(f"   Model: {model_name}")
print(f"   Max sequence length: {max_seq_length}")
print(f"   4-bit quantization: {load_in_4bit}")
print("\n   This will take 3-5 minutes (downloading ~8GB)...\n")

# Load model with Unsloth
model, tokenizer = FastLanguageModel.from_pretrained(
    model_name=model_name,
    max_seq_length=max_seq_length,
    dtype=dtype,
    load_in_4bit=load_in_4bit,
)

print("\n   Model loaded successfully!")
print(f"   Model parameters: {model.num_parameters():,}")
print(f"   Memory allocated: {torch.cuda.memory_allocated() / 1e9:.2f} GB")

---

# Step 4: Add LoRA Adapters

## Understanding LoRA (Low-Rank Adaptation)

### The Problem:
Traditional fine-tuning updates **ALL 4 billion parameters**:
- Requires massive memory (store gradients for 4B params)
- Very slow (update 4B weights)
- Easy to overfit on small datasets

### LoRA Solution:
Instead of modifying original weights, add **small adapter matrices**:

```
Original Weight Matrix W (large):
[2048 × 2048] = 4,194,304 parameters

LoRA Decomposition:
ΔW = A × B
A: [2048 × 16]  = 32,768 parameters
B: [16 × 2048]  = 32,768 parameters
Total: 65,536 parameters (64x smaller!)

Final Output:
y = W·x + α·(A·B)·x
    ↑      ↑
 frozen  trainable
```

### Key Parameters:

1. **r (rank)**: Size of adapter matrices (typically 8-16)
   - Higher r = more capacity but slower
   - Lower r = faster but less expressive
   - We use r=16 (good balance)

2. **lora_alpha**: Scaling factor for LoRA updates
   - Controls how much LoRA affects output
   - Typically 2×r (we use 32)

3. **lora_dropout**: Regularization (prevent overfitting)
   - We use 0 (dataset is diverse enough)

4. **target_modules**: Which layers to adapt
   - `q_proj`, `k_proj`: Query/Key attention projections
   - `v_proj`, `o_proj`: Value/Output projections
   - `gate_proj`, `up_proj`, `down_proj`: MLP layers

### Memory Savings:
```
Without LoRA: 4B params × 2 bytes = 8 GB
With LoRA: ~50M params × 2 bytes = 100 MB

Savings: ~98.7% reduction in trainable parameters!
```

In [8]:
# Add LoRA adapters to the model
# These are small matrices we'll train instead of the full model

model = FastLanguageModel.get_peft_model(
    model,
    r=16,  # LoRA rank (size of adapter matrices)
    target_modules=[
        "q_proj",     # Query projection in attention
        "k_proj",     # Key projection
        "v_proj",     # Value projection
        "o_proj",     # Output projection
        "gate_proj",  # MLP gate
        "up_proj",    # MLP up
        "down_proj",  # MLP down
    ],
    lora_alpha=32,  # LoRA scaling factor (typically 2×r)
    lora_dropout=0,  # No dropout (our dataset is diverse)
    bias="none",     # Don't train bias terms
    use_gradient_checkpointing="unsloth",  # Memory optimization
    random_state=42,  # Reproducibility
    use_rslora=False,  # Standard LoRA (RSLoRA is for very large models)
    loftq_config=None,  # No LoftQ quantization
)

print("✅ LoRA adapters added!")
print("\n📊 Model Statistics:")

# Count trainable vs frozen parameters
trainable_params = sum(p.numel() for p in model.parameters() if p.requires_grad)
total_params = sum(p.numel() for p in model.parameters())
trainable_pct = 100 * trainable_params / total_params

print(f"   Total parameters: {total_params:,}")
print(f"   Trainable parameters: {trainable_params:,}")
print(f"   Trainable %: {trainable_pct:.2f}%")
print(f"   Memory allocated: {torch.cuda.memory_allocated() / 1e9:.2f} GB")

print("\n💡 Insight:")
print(f"   We're training only {trainable_pct:.2f}% of parameters!")
print(f"   This is why LoRA is so efficient.")

Unsloth: Making `model.base_model.model.model` require gradients
✅ LoRA adapters added!

📊 Model Statistics:
   Total parameters: 675,994,752
   Trainable parameters: 13,045,760
   Trainable %: 1.93%
   Memory allocated: 1.05 GB

💡 Insight:
   We're training only 1.93% of parameters!
   This is why LoRA is so efficient.


---

# Step 5: Prepare Dataset for Training

## 📖 Understanding the Training Format

### Alpaca Format:
Our dataset uses the Alpaca instruction format:
```json
{
  "instruction": "Task description",
  "input": "Additional context (empty for us)",
  "output": "Expected response"
}
```

### How it's converted for training:
```
Alpaca Format:
  instruction: "Translate to venvy command: list all environments"
  input: ""
  output: "COMMAND: venvy ls\nCONFIDENCE: 0.95\n..."

↓ Transformed to ↓

Gemma 3 Chat Format:
<start_of_turn>user
Translate to venvy command: list all environments<end_of_turn>
<start_of_turn>model
COMMAND: venvy ls
CONFIDENCE: 0.95
EXPLANATION: Lists all registered virtual environments
<end_of_turn>
```

### Why this format?
- Gemma 3 is trained as a chat model with turn-based conversation
- `<start_of_turn>user` signals user input
- `<start_of_turn>model` signals model response
- This matches how Gemma 3 was pre-trained

In [ ]:
from datasets import load_dataset

# Load dataset from JSONL file
dataset = load_dataset('json', data_files=DATASET_FILE, split='train')

print(f"Dataset loaded: {len(dataset)} examples")
print(f"\nDataset Structure:")
print(dataset)
print(f"\nSample Example:")
print(dataset[0])

In [10]:
# Split dataset: 90% train, 10% validation
# Validation set helps us monitor if the model is overfitting

dataset = dataset.train_test_split(test_size=0.1, seed=42)
train_dataset = dataset['train']
eval_dataset = dataset['test']

print(f"✅ Dataset split:")
print(f"   Training examples: {len(train_dataset)}")
print(f"   Validation examples: {len(eval_dataset)}")

print("\n💡 Why validation set?")
print("   We'll evaluate on this during training to detect overfitting.")
print("   If validation loss stops improving, we stop training.")

✅ Dataset split:
   Training examples: 1350
   Validation examples: 150

💡 Why validation set?
   We'll evaluate on this during training to detect overfitting.
   If validation loss stops improving, we stop training.


In [11]:
# Format dataset for Gemma 3 chat format
# This converts our Alpaca format to Gemma's expected input format

# Gemma 3 chat template
alpaca_prompt = """<start_of_turn>user
{}<end_of_turn>
<start_of_turn>model
{}<end_of_turn>"""

EOS_TOKEN = tokenizer.eos_token  # End-of-sequence token

def formatting_prompts_func(examples):
    """
    Convert Alpaca format to Gemma 3 chat format.

    For each example:
    1. Combine instruction + input (input is empty for us)
    2. Format as Gemma chat turn
    3. Add EOS token for proper training
    """
    instructions = examples["instruction"]
    inputs = examples["input"]
    outputs = examples["output"]

    texts = []
    for instruction, input_text, output in zip(instructions, inputs, outputs):
        # Combine instruction and input (input is empty for our dataset)
        full_instruction = instruction + ("\n" + input_text if input_text else "")

        # Format as chat turns
        text = alpaca_prompt.format(full_instruction, output) + EOS_TOKEN
        texts.append(text)

    return {"text": texts}

# Apply formatting to both train and validation sets
train_dataset = train_dataset.map(
    formatting_prompts_func,
    batched=True,
)

eval_dataset = eval_dataset.map(
    formatting_prompts_func,
    batched=True,
)

print("✅ Dataset formatted for Gemma 3 chat!")
print("\n📝 Formatted Example:")
print("-" * 80)
print(train_dataset[0]['text'])
print("-" * 80)

Map:   0%|          | 0/1350 [00:00<?, ? examples/s]

Map:   0%|          | 0/150 [00:00<?, ? examples/s]

✅ Dataset formatted for Gemma 3 chat!

📝 Formatted Example:
--------------------------------------------------------------------------------
<start_of_turn>user
Translate to venvy command: preview cleanup<end_of_turn>
<start_of_turn>model
COMMAND: venvy cleanup --dry-run
CONFIDENCE: 0.97
EXPLANATION: Shows which environments would be removed without deleting
<end_of_turn><end_of_turn>
--------------------------------------------------------------------------------


---

# Step 6: Configure Training Parameters

## Understanding Hyperparameters

### Key Training Parameters:

#### 1. **Learning Rate (lr)**: How fast the model learns
```
Too high (1e-3):  Model diverges, loss explodes
Just right (2e-4): Smooth learning, converges well
Too low (1e-5):   Learns too slowly, wastes time
```
We use **2e-4** (0.0002) - standard for LoRA fine-tuning.

#### 2. **Batch Size**: How many examples per update
```
per_device_batch_size=2:  Process 2 examples at once (reduced for 4B model)
gradient_accumulation_steps=8: Accumulate 8 batches
Effective batch size = 2 × 8 = 16
```
Why smaller per-device batch?
- 4B model uses ~2.5GB VRAM in 4-bit
- LoRA + optimizer + activations need ~8GB more
- Batch size 2 keeps us safely within T4's 16GB VRAM
- We compensate with more gradient accumulation steps

#### 3. **Epochs**: How many times to see full dataset
```
1 epoch = model sees each example once
3 epochs = model sees each example 3 times
```
We use **3 epochs** - enough to learn without overfitting.

#### 4. **Weight Decay**: Regularization to prevent overfitting
```
weight_decay=0.01: Small penalty on large weights
```
Encourages model to use many small weights instead of few large ones.

#### 5. **Learning Rate Schedule**: Warmup + Cosine Decay
```
Step 0-50:    Warmup (gradual increase) → Prevents early instability
Step 50-end:  Cosine decay (gradual decrease) → Better convergence

Learning Rate over time:
    |
2e-4|        _______________
    |      /                 \
    |    /                     \
    |  /                         \
  0 |_/____________________________\_____
     0   50                   1000  steps
```

#### 6. **Mixed Precision (FP16)**: Speed + Memory optimization
```
Normal (FP32):  32 bits per number → Slow but accurate
Mixed (FP16):   16 bits per number → 2x faster, 2x less memory
```
T4 GPU has FP16 cores (Tensor Cores) → much faster.

### Expected Training Time:
```
594 examples × 3 epochs / (batch_size 16) = ~111 update steps
~3-5 seconds per step on T4 (4B is larger)
Total: ~30-45 minutes
```

In [ ]:
from transformers import TrainingArguments
from trl import SFTTrainer

# Training configuration
# NOTE: batch_size reduced from 4→2 for 4B model on T4 (16GB VRAM)
# gradient_accumulation increased from 4→8 to keep effective batch=16
training_args = TrainingArguments(
    # Output and logging
    output_dir="./outputs",              # Where to save model checkpoints
    logging_dir="./logs",                # Where to save logs
    logging_steps=10,                    # Log every 10 steps

    # Training hyperparameters
    num_train_epochs=3,                  # Train for 3 epochs
    per_device_train_batch_size=2,       # 2 examples per GPU (reduced for 4B)
    gradient_accumulation_steps=8,       # Accumulate 8 batches (effective batch=16)
    learning_rate=2e-4,                  # Standard LoRA learning rate
    weight_decay=0.01,                   # L2 regularization

    # Learning rate schedule
    lr_scheduler_type="cosine",          # Cosine decay schedule
    warmup_steps=50,                     # Warmup for first 50 steps

    # Optimization
    optim="adamw_8bit",                  # 8-bit AdamW (saves memory)
    fp16=True,                           # Mixed precision training (2x faster)

    # Evaluation
    eval_strategy="steps",               # Evaluate during training
    eval_steps=50,                       # Evaluate every 50 steps
    per_device_eval_batch_size=2,        # Batch size for evaluation

    # Checkpointing
    save_strategy="steps",               # Save checkpoints
    save_steps=100,                      # Save every 100 steps
    save_total_limit=3,                  # Keep only 3 best checkpoints
    load_best_model_at_end=True,         # Load best checkpoint at end
    metric_for_best_model="eval_loss",   # Use validation loss to pick best

    # Memory optimizations
    gradient_checkpointing=True,         # Save memory (slight speed cost)
    max_grad_norm=1.0,                   # Gradient clipping (stability)

    # Reproducibility
    seed=42,

    # Disable unnecessary features
    report_to="none",                    # Don't report to wandb/tensorboard
)

print("Training configuration set!")
print("\nTraining Summary:")
print(f"   Epochs: {training_args.num_train_epochs}")
print(f"   Per-device batch size: {training_args.per_device_train_batch_size}")
print(f"   Gradient accumulation: {training_args.gradient_accumulation_steps}")
print(f"   Effective batch size: {training_args.per_device_train_batch_size * training_args.gradient_accumulation_steps}")
print(f"   Learning rate: {training_args.learning_rate}")
print(f"   Warmup steps: {training_args.warmup_steps}")
print(f"   FP16 enabled: {training_args.fp16}")

# Calculate approximate training time
total_steps = (len(train_dataset) * training_args.num_train_epochs) // (training_args.per_device_train_batch_size * training_args.gradient_accumulation_steps)
print(f"\nEstimated training time:")
print(f"   Total steps: ~{total_steps}")
print(f"   Time: ~{total_steps * 4 / 60:.1f} minutes (assuming ~4 sec/step for 4B)")

---

# Step 7: Train the Model!

## What Happens During Training?

### Training Loop:
```python
for epoch in range(3):
    for batch in train_dataset:
        # 1. Forward pass: Get model predictions
        predictions = model(batch)
        
        # 2. Calculate loss: How wrong are we?
        loss = cross_entropy(predictions, targets)
        
        # 3. Backward pass: Calculate gradients
        gradients = loss.backward()
        
        # 4. Update LoRA weights
        optimizer.step(gradients)
        
        # 5. Log progress
        if step % 10 == 0:
            print(f"Loss: {loss:.4f}")
```

### What to Watch:

1. **Training Loss**: Should decrease smoothly
   ```
   Good:    2.5 → 1.8 → 1.2 → 0.8 → 0.5
   Bad:     2.5 → 5.8 → NaN (model diverged!)
   ```

2. **Validation Loss**: Should also decrease
   ```
   Good:    Train loss ≈ Val loss (not overfitting)
   Bad:     Train 0.3, Val 2.5 (overfitting!)
   ```

3. **Speed**: Should be ~3-5 seconds per step for 4B model
   - Slower than 1B (~1-2s) due to 4x more parameters
   - If OOM: reduce batch size or restart runtime

### Training Metrics Explained:

- **loss**: Cross-entropy loss (lower = better)
- **learning_rate**: Current LR (starts low, increases, then decreases)
- **epoch**: Which epoch we're on (0-3)
- **grad_norm**: Gradient magnitude (should be stable, not exploding)

This cell will take ~30-45 minutes on T4. Grab a coffee!

In [13]:
# Create trainer with SFTTrainer (Supervised Fine-Tuning)
trainer = SFTTrainer(
    model=model,
    tokenizer=tokenizer,
    train_dataset=train_dataset,
    eval_dataset=eval_dataset,
    dataset_text_field="text",  # Which field contains the formatted text
    max_seq_length=max_seq_length,
    args=training_args,
    packing=False,  # Don't pack multiple examples (our examples are short)
)

print("✅ Trainer initialized!")
print("\n🚀 Starting training...")
print("   This will take ~8-10 minutes on T4 GPU")
print("   Watch the loss decrease over time!")
print("\n" + "="*80)

Unsloth: Switching to float32 training since model cannot work with float16


Unsloth: Tokenizing ["text"] (num_proc=6):   0%|          | 0/1350 [00:00<?, ? examples/s]

Unsloth: Tokenizing ["text"] (num_proc=6):   0%|          | 0/150 [00:00<?, ? examples/s]

✅ Trainer initialized!

🚀 Starting training...
   This will take ~8-10 minutes on T4 GPU
   Watch the loss decrease over time!



In [14]:
# Start training!
# The output will show:
# - Loss (should decrease)
# - Learning rate (should follow warmup + cosine schedule)
# - Time per step
# - Memory usage

trainer_stats = trainer.train()

print("\n" + "="*80)
print("🎉 Training complete!")
print("\n📊 Final Statistics:")
print(f"   Train runtime: {trainer_stats.metrics['train_runtime']:.2f} seconds")
print(f"   Train samples/second: {trainer_stats.metrics['train_samples_per_second']:.2f}")
print(f"   Final train loss: {trainer_stats.metrics['train_loss']:.4f}")

# Get validation metrics
eval_results = trainer.evaluate()
print(f"\n📈 Validation Results:")
print(f"   Validation loss: {eval_results['eval_loss']:.4f}")
print(f"   Validation perplexity: {eval_results.get('eval_perplexity', 'N/A')}")

==((====))==  Unsloth - 2x faster free finetuning | Num GPUs used = 1
   \\   /|    Num examples = 1,350 | Num Epochs = 3 | Total steps = 255
O^O/ \_/ \    Batch size per device = 4 | Gradient accumulation steps = 4
\        /    Data Parallel GPUs = 1 | Total batch size (4 x 4 x 1) = 16
 "-____-"     Trainable parameters = 13,045,760 of 1,012,931,712 (1.29% trained)


Step,Training Loss,Validation Loss
50,0.497600,0.368994
100,0.193300,0.181701
150,0.155700,0.148017
200,0.139300,0.142402
250,0.135100,0.138184


Unsloth: Not an error, but Gemma3ForCausalLM does not accept `num_items_in_batch`.
Using gradient accumulation will be very slightly less accurate.
Read more on gradient accumulation issues here: https://unsloth.ai/blog/gradient



🎉 Training complete!

📊 Final Statistics:
   Train runtime: 617.72 seconds
   Train samples/second: 6.56
   Final train loss: 0.6653



📈 Validation Results:
   Validation loss: 0.1424
   Validation perplexity: N/A


---

# Step 8: Test the Model

Let's see if our fine-tuned model can actually translate natural language to venvy commands!

In [ ]:
# Enable inference mode (faster, less memory)
FastLanguageModel.for_inference(model)

def test_command_translation(nl_query):
    """Test the model's ability to translate natural language to CLI commands."""
    instruction = f"Translate to {CLI_TOOL} command: {nl_query}"
    prompt = alpaca_prompt.format(instruction, "")
    inputs = tokenizer(text=prompt, return_tensors="pt").to("cuda")
    outputs = model.generate(
        **inputs,
        max_new_tokens=128,
        temperature=0.1,
        top_p=0.9,
        do_sample=True,
    )
    response = tokenizer.decode(outputs[0], skip_special_tokens=True)
    if "<start_of_turn>model" in response:
        response = response.split("<start_of_turn>model")[-1].strip()
    return response

print("Inference mode enabled!")
print(f"\nTesting model on {CLI_TOOL} queries...\n")
print("="*80)

In [ ]:
# Simple test
prompt = "<start_of_turn>user\nHello<end_of_turn>\n<start_of_turn>model\n"
inputs = tokenizer(text=prompt, return_tensors="pt").to("cuda")

outputs = model.generate(
    **inputs,
    max_new_tokens=50,
    do_sample=False,
)

print("Full output:")
print(tokenizer.decode(outputs[0], skip_special_tokens=False))
print("\n" + "="*80 + "\n")
print("Without special tokens:")
print(tokenizer.decode(outputs[0], skip_special_tokens=True))

In [18]:
# Verify how training data looked
print("Sample training example:")
print(train_dataset[0]['text'])

Sample training example:
<start_of_turn>user
Translate to venvy command: preview cleanup<end_of_turn>
<start_of_turn>model
COMMAND: venvy cleanup --dry-run
CONFIDENCE: 0.97
EXPLANATION: Shows which environments would be removed without deleting
<end_of_turn><end_of_turn>


In [ ]:
# Use EXACT format from training
prompt = f"""<start_of_turn>user
Translate to {CLI_TOOL} command: show all running containers<end_of_turn>
<start_of_turn>model
"""

inputs = tokenizer(text=prompt, return_tensors="pt").to("cuda")

outputs = model.generate(
    **inputs,
    max_new_tokens=128,
    do_sample=False,
    pad_token_id=tokenizer.eos_token_id,
)

full_output = tokenizer.decode(outputs[0], skip_special_tokens=False)
print("Full output:")
print(full_output)

In [ ]:

def test_command_translation_FIXED(nl_query):
    instruction = f"Translate to {CLI_TOOL} command: {nl_query}"
    prompt = f"<start_of_turn>user\n{instruction}<end_of_turn>\n<start_of_turn>model\n"

    inputs = tokenizer(text=prompt, return_tensors="pt").to("cuda")
    outputs = model.generate(
        **inputs,
        max_new_tokens=128,
        temperature=0.1,
        do_sample=True,
        pad_token_id=tokenizer.eos_token_id,
    )

    full_response = tokenizer.decode(outputs[0], skip_special_tokens=False)

    if "<start_of_turn>model\n" in full_response:
        response = full_response.split("<start_of_turn>model\n")[-1]
        if "<end_of_turn>" in response:
            response = response.split("<end_of_turn>")[0]
        return response.strip()

    return full_response

# Test it
queries = [
    "show all running containers",
    "run nginx on port 8080 in background",
    "build image tagged myapp version 2",
    "scale web service to 3 instances",
    "show logs for container api",
    "list docker networks",
]

print(f"TESTING {CLI_TOOL.upper()} MODEL:")
print("="*80)
for q in queries:
    result = test_command_translation_FIXED(q)
    print(f"\nQuery: {q}")
    print(f"Output:\n{result}")
    print("-"*80)

---

# Step 9: Save the Fine-tuned Model

We'll save both:
1. **LoRA adapters only** (small, ~16MB)
2. **Merged model** (base + adapters, ~2.2GB)

In [ ]:
# Save LoRA adapters only (small file, quick to upload/download)
lora_dir = f"{OUTPUT_PREFIX}_lora"
model.save_pretrained(lora_dir)
tokenizer.save_pretrained(lora_dir)

print(f"LoRA adapters saved to: {lora_dir}/")
print("   Size: ~16MB (adapters only)")
print(f"\nTo load later:")
print(f"   model = FastLanguageModel.from_pretrained('{lora_dir}')")

In [ ]:
# Merge LoRA adapters into base model (for GGUF conversion)
merged_dir = f"{OUTPUT_PREFIX}_merged"
print(f"Merging LoRA adapters into base model...")
print("   This combines the base Gemma 3 1B with our trained adapters")
print("   Result will be ~2.2GB in FP16 format")

model.save_pretrained_merged(
    merged_dir,
    tokenizer,
    save_method="merged_16bit",  # Save in FP16 (2 bytes per param)
)

print(f"\nMerged model saved to: {merged_dir}/")
print("   Size: ~2.2GB (full model in FP16)")
print("\nNext step: Convert to GGUF for CPU inference")

---

# Step 10: Convert to GGUF Format

## Understanding GGUF Conversion

### Why GGUF?
**GGUF** (GPT-Generated Unified Format) is optimized for CPU inference:
- Used by llama.cpp for efficient CPU/Metal/Vulkan inference
- Supports various quantization levels (2-bit to 8-bit)
- Memory-mapped for fast loading
- Cross-platform (Windows, Mac, Linux)

### Quantization Options (4B model):
```
Q2_K: 2-bit → ~1.2GB, fast but lower quality
Q3_K_M: 3-bit → ~1.7GB, good balance
Q4_0: 4-bit basic → ~2.1GB, standard
Q4_K_M: 4-bit with K-means → ~2.5GB, better quality  (our choice)
Q5_K_M: 5-bit with K-means → ~3.0GB, excellent quality
Q8_0: 8-bit → ~4.2GB, minimal loss
```

### K-means Quantization:
Instead of uniform quantization, K-means clusters weights:
```
Standard Q4: [-1.0, -0.5, 0.0, 0.5, 1.0] (uniform bins)
K-means Q4:  [-0.9, -0.3, 0.1, 0.6, 1.2] (optimized bins)
                                         ↑
                        Better matches weight distribution
```

### Importance Matrix (imatrix):
Identifies which layers are most important for your specific task:
1. Run inference on your dataset
2. Measure activation magnitudes per layer
3. Quantize unimportant layers more aggressively
4. Preserve critical layers with higher precision

Result: **15-20% better quality** at same size

In [56]:
%%capture
!git clone https://github.com/ggerganov/llama.cpp

In [57]:

print("🔨 Building llama.cpp with CMake...")
!mkdir -p llama.cpp/build
!cd llama.cpp/build && cmake .. -DCMAKE_BUILD_TYPE=Release
!cd llama.cpp/build && cmake --build . --config Release --target llama-quantize llama-imatrix -j 4

🔨 Building llama.cpp with CMake...
CMAKE_BUILD_TYPE=Release
-- Warning: ccache not found - consider installing it for faster compilation or disable this warning with GGML_CCACHE=OFF
-- CMAKE_SYSTEM_PROCESSOR: x86_64
-- GGML_SYSTEM_ARCH: x86
-- Including CPU backend
-- x86 detected
-- Adding CPU backend variant ggml-cpu: -march=native 
-- ggml version: 0.9.4
-- ggml commit:  eeee367de
-- Configuring done (0.7s)
-- Generating done (0.5s)
-- Build files have been written to: /content/nlcli-wizard/llama.cpp/build
[  0%] Built target build_info
[  4%] Built target ggml-base
[ 14%] Built target ggml-cpu
[ 17%] Built target ggml
[ 17%] Building CXX object src/CMakeFiles/llama.dir/unicode.cpp.o
[ 17%] Building CXX object src/CMakeFiles/llama.dir/llama-model.cpp.o
[ 17%] Building CXX object src/CMakeFiles/llama.dir/models/deepseek2.cpp.o
[ 19%] Building CXX object src/CMakeFiles/llama.dir/models/dots1.cpp.o
[ 19%] Building CXX object src/CMakeFiles/llama.dir/models/dream.cpp.o
[ 19%] Building C

In [58]:
print("✅ llama.cpp built!")

✅ llama.cpp built!


In [54]:
# Install llama.cpp for GGUF conversion
# %%capture
# !git clone https://github.com/ggerganov/llama.cpp
# !cd llama.cpp && make

# print("✅ llama.cpp installed!")


In [ ]:
# Step 1: Convert HuggingFace model to GGUF FP16
fp16_file = f"{OUTPUT_PREFIX}_fp16.gguf"
print(f"Step 1: Converting to GGUF FP16 format...")

!python llama.cpp/convert_hf_to_gguf.py \
    {merged_dir} \
    --outfile {fp16_file} \
    --outtype f16

print(f"\nGGUF FP16 model created: {fp16_file} (~2.2GB)")

In [ ]:
# ============================================================
# STEP 2: Generate Importance Matrix (CMAKE BUILD - 2025)
# ============================================================

imatrix_file = f"{OUTPUT_PREFIX}_imatrix.dat"

print("Step 2: Generating importance matrix...")
print("\nWhat's an importance matrix?")
print(f"   Identifies which model layers are CRITICAL for {CLI_TOOL} commands")
print("   vs less important layers we can compress more aggressively.")
print("\nTakes ~5-10 minutes...\n")

# Create a text file with sample commands for imatrix generation
import json
import os

with open('imatrix_data.txt', 'w') as f:
    for i, example in enumerate(train_dataset.select(range(min(100, len(train_dataset))))):
        f.write(example['text'] + '\n\n')

print(f"Created imatrix_data.txt with 100 {CLI_TOOL} examples")

# Build llama.cpp with CMAKE (new build system as of 2025)
print("\nStep 2a: Building llama.cpp with CMake...")

if not os.path.exists('llama.cpp/build/bin/llama-imatrix'):
    print("   Installing build tools...")
    !apt-get update -qq
    !apt-get install -y -qq cmake build-essential

    print("   Building llama-imatrix (takes ~2-3 minutes)...")
    !mkdir -p llama.cpp/build
    !cd llama.cpp/build && cmake .. -DCMAKE_BUILD_TYPE=Release
    !cd llama.cpp/build && cmake --build . --config Release --target llama-imatrix -j 4

    print("Build complete!")
else:
    print("llama-imatrix already built!")

# Verify build
if os.path.exists('llama.cpp/build/bin/llama-imatrix'):
    print(f"Tool verified: llama.cpp/build/bin/llama-imatrix")

    # Generate importance matrix
    print(f"\nStep 2b: Running importance analysis...")
    print("   Processing 100 examples to measure layer activations...")

    !llama.cpp/build/bin/llama-imatrix \
        -m {fp16_file} \
        -f imatrix_data.txt \
        -o {imatrix_file} \
        --chunks 100 \
        -ngl 0 \
        -t 4

    if os.path.exists(imatrix_file) and os.path.getsize(imatrix_file) > 0:
        print(f"\nImportance matrix generated: {imatrix_file}")
    else:
        print("\nimatrix generation failed, creating dummy file")
        !touch {imatrix_file}
else:
    print("\nBuild failed - skipping imatrix (model will still be good!)")
    !touch {imatrix_file}

In [ ]:
# ============================================================
# STEP 3: Quantize to Q4_K_M
# ============================================================

q4km_file = f"{OUTPUT_PREFIX}_q4km.gguf"

print("Step 3: Quantizing to Q4_K_M with importance matrix...")

import os

# Check if imatrix file exists
if os.path.exists(imatrix_file):
    imatrix_size = os.path.getsize(imatrix_file) / 1024
    print(f"\nimatrix file exists: {imatrix_size:.2f} KB")

print("\nAttempting quantization...\n")

# Try with imatrix first
try:
    !llama.cpp/build/bin/llama-quantize \
        {fp16_file} \
        {q4km_file} \
        Q4_K_M \
        --imatrix {imatrix_file}

    if os.path.exists(q4km_file) and os.path.getsize(q4km_file) > 0:
        print("Method 1 (with imatrix) worked!")
    else:
        raise Exception("Method 1 failed")
except:
    print("Method 1 (--imatrix) didn't work, trying without...")

    !llama.cpp/build/bin/llama-quantize \
        {fp16_file} \
        {q4km_file} \
        Q4_K_M

# Verify result
if os.path.exists(q4km_file) and os.path.getsize(q4km_file) > 0:
    print("\nQuantized model created!")

    fp16_size = os.path.getsize(fp16_file) / 1e9
    q4km_size = os.path.getsize(q4km_file) / 1e9
    compression_ratio = fp16_size / q4km_size

    print(f"\nCompression Statistics:")
    print(f"   Original (FP16): {fp16_size:.2f} GB")
    print(f"   Quantized (Q4_K_M): {q4km_size:.2f} GB")
    print(f"   Compression ratio: {compression_ratio:.1f}x")
    print(f"   Space saved: {(fp16_size - q4km_size):.2f} GB")
else:
    print("\nQuantization failed!")

---

# Step 11: Test GGUF Model

Let's verify the quantized model works correctly!

In [62]:
# Install llama-cpp-python for testing
%%capture
!pip install llama-cpp-python

print("✅ llama-cpp-python installed!")

In [ ]:
from llama_cpp import Llama

# Load GGUF model
print("Loading GGUF model...")

llm = Llama(
    model_path=q4km_file,
    n_ctx=512,  # Context window
    n_threads=4,  # CPU threads
    verbose=False,
)

print(f"GGUF model loaded!")
print(f"   Model: {q4km_file}")
print(f"   Size: {q4km_size:.2f} GB")
print(f"   Context window: 512 tokens")

In [ ]:
# Test GGUF model on Docker queries
def test_gguf_translation(nl_query):
    instruction = f"Translate to {CLI_TOOL} command: {nl_query}"
    prompt = f"<start_of_turn>user\n{instruction}<end_of_turn>\n<start_of_turn>model\n"

    response = llm(
        prompt,
        max_tokens=128,
        temperature=0.1,
        top_p=0.9,
        stop=["<end_of_turn>"],
    )

    return response['choices'][0]['text'].strip()

print(f"Testing GGUF {CLI_TOOL} model...\n")
print("="*80)

test_queries = [
    "show all running containers",
    "run nginx on port 8080 in background",
    "build image tagged myapp version 2",
    "scale web service to 3 instances",
    "show logs for container api",
    "list docker networks",
]

for query in test_queries:
    response = test_gguf_translation(query)
    print(f"Query: {query}")
    print(f"Response: {response}")
    print("-"*80)

print(f"\nIf the responses look correct, your {CLI_TOOL} GGUF model is ready!")

---

# Step 12: Download the Models

Download these files to your local machine:
1. `docker_gemma3_4b_q4km.gguf` - Final quantized model (~2.5GB)
2. `docker_gemma3_4b_lora/` - LoRA adapters (~50MB)

**Tip**: If `files.download()` doesn't work for the large GGUF file,
use the Google Drive workaround:
```python
from google.colab import drive
import shutil
drive.mount('/content/drive')
shutil.copy(q4km_file, f'/content/drive/MyDrive/{q4km_file}')
```

In [ ]:
# Download the quantized model
from google.colab import files

print(f"Downloading {q4km_file}...")
print("   This may take a few minutes for the ~800MB file\n")

files.download(q4km_file)

print(f"\nModel downloaded!")
print(f"\nNext steps:")
print(f"   1. Place {q4km_file} in models/ directory of nlcli-wizard")
print(f"   2. Run: python test/evaluate_docker.py")
print(f"   3. Run: python -m nlcli_wizard.cli translate --cli-tool docker show running containers")

---

# Training Complete!

## What You Accomplished:

1. Fine-tuned Gemma 3 4B on 594 Docker command examples
2. Used QLoRA for efficient training (~1.3% parameter reduction)
3. Leveraged Unsloth for 2x speed, 70% less VRAM
4. Quantized to Q4_K_M with importance matrix
5. Created GGUF model for CPU inference (~2.5GB)

## 1B vs 4B Comparison:

| Metric | Gemma 3 1B | Gemma 3 4B |
|--------|-----------|-----------|
| Accuracy | 73-76% | TBD |
| Model Size | ~800MB | ~2.5GB |
| Inference | ~5s | ~8-12s |
| Training | ~16 min | ~30-45 min |
| Trainable Params | 13M (1.29%) | ~50M (~1.3%) |

## Files to Keep:

```
docker_gemma3_4b_q4km.gguf        # Final model (~2.5GB) - THIS IS WHAT YOU NEED
docker_gemma3_4b_lora/             # LoRA adapters (~50MB)
docker_gemma3_4b_imatrix.dat       # Importance matrix
```

## Next Steps:

1. Place `docker_gemma3_4b_q4km.gguf` in `models/` directory
2. Run `python test/evaluate_docker.py` for accuracy metrics
3. Run `python -m nlcli_wizard.cli list-tools` to verify
4. Test: `python -m nlcli_wizard.cli translate --cli-tool docker show running containers`
5. Post 1B vs 4B comparison results to Reddit/LinkedIn

---

**Congratulations! You've trained a 4B Docker command translator!**

In [ ]:
# test_docker.py - Quick test script

from llama_cpp import Llama

# Load model
llm = Llama(
    model_path="models/docker_gemma3_4b_q4km.gguf",
    n_ctx=512,
    n_threads=4,
    verbose=False,
)

def translate(nl_query):
    """Translate natural language to docker command."""
    prompt = f"""<start_of_turn>user
Translate to docker command: {nl_query}<end_of_turn>
<start_of_turn>model
"""
    response = llm(prompt, max_tokens=128, temperature=0.1, stop=["<end_of_turn>"])
    return response['choices'][0]['text'].strip()

# Demo
queries = [
    "show all running containers",
    "run nginx on port 8080 in background",
    "build image tagged myapp",
    "show logs for container web",
    "scale web service to 3 instances",
    "list docker networks",
]

print("Gemma 3 4B - Docker Command Translator\n")
print("="*80)

for query in queries:
    result = translate(query)
    print(f"\nQuery: \"{query}\"")
    print(f"Output:")
    print(result)
    print("-"*80)

In [ ]:
# evaluate_docker.py - Accuracy evaluation with per-category breakdown
from llama_cpp import Llama
import json
import time

llm = Llama(model_path="models/docker_gemma3_4b_q4km.gguf", n_ctx=512, n_threads=4, verbose=False)

with open('data/docker_training.jsonl') as f:
    examples = [json.loads(line) for line in f][-100:]

correct = 0
total = len(examples)
category_stats = {}
inference_times = []

for ex in examples:
    query = ex['instruction'].replace('Translate to docker command: ', '')
    expected_cmd = ex['output'].split('COMMAND: ')[1].split('\n')[0]

    if expected_cmd.startswith('docker-compose'):
        category = 'compose'
    elif expected_cmd.startswith('docker run'):
        category = 'run'
    elif expected_cmd.startswith('docker build'):
        category = 'build'
    elif expected_cmd.startswith('docker exec'):
        category = 'exec'
    elif expected_cmd.startswith('docker network'):
        category = 'network'
    elif expected_cmd.startswith('docker volume'):
        category = 'volume'
    elif expected_cmd.startswith(('docker system', 'docker info', 'docker version',
                                  'docker container', 'docker image')):
        category = 'system'
    else:
        category = 'ps_images'

    if category not in category_stats:
        category_stats[category] = {'correct': 0, 'total': 0}
    category_stats[category]['total'] += 1

    prompt = f"<start_of_turn>user\n{ex['instruction']}<end_of_turn>\n<start_of_turn>model\n"
    start = time.time()
    response = llm(prompt, max_tokens=128, temperature=0.1, stop=["<end_of_turn>"])
    inference_times.append(time.time() - start)
    predicted = response['choices'][0]['text'].strip()

    if 'COMMAND:' in predicted:
        predicted_cmd = predicted.split('COMMAND: ')[1].split('\n')[0].strip()
    else:
        predicted_cmd = predicted.split('\n')[0].strip()

    if predicted_cmd == expected_cmd:
        correct += 1
        category_stats[category]['correct'] += 1
    else:
        print(f"MISS | {category:10} | Query: {query}")
        print(f"     | Expected:  {expected_cmd}")
        print(f"     | Got:       {predicted_cmd}\n")

accuracy = correct / total
avg_time = sum(inference_times) / len(inference_times)

print(f"\n{'='*60}")
print(f"DOCKER MODEL EVALUATION RESULTS (4B)")
print(f"{'='*60}")
print(f"Overall Accuracy: {correct}/{total} = {accuracy:.1%}")
print(f"Avg Inference:    {avg_time:.2f}s per query")
print(f"{'='*60}")
print(f"\nPer-category breakdown:")
for cat in sorted(category_stats.keys()):
    stats = category_stats[cat]
    cat_acc = stats['correct'] / stats['total'] if stats['total'] > 0 else 0
    print(f"  {cat:12} {stats['correct']:>3}/{stats['total']:<3} = {cat_acc:.1%}")